### PR TITLE
Refactor Nosnoc Problem Setup

### PR DESCRIPTION
**Pull request by FreyJo**

### CODE REVIEW

**Review by apozharski — COMMENTED** [review]
I'm thinking it might pay off long term to have a stage object, if we will also later have cross-complementarity modes that sum across stages. Not necessary for now.

Aside from that, just a couple small things.

File: nosnoc/nosnoc.py [hunk 1]
```diff
@@ -65,16 +65,21 @@ class NosnocDims:
...
         self.w = SX([])
         self.w0 = np.array([])
         self.lbw = np.array([])
         self.ubw = np.array([])
 
+        self.g = SX([])
+        self.lbg = np.array([])
+        self.ubg = np.array([])
+
+        self.cost: float = 0.0
```
[the remark this answers is not in the document]
Comment (FreyJo) [reply nosnoc/nosnoc.py:81]: Done!

File: nosnoc/nosnoc.py [hunk 2]
```diff
@@ -626,192 +713,65 @@ def __init__(self, settings: NosnocSettings, model: NosnocModel, ocp=None):
...
-                            nu_k / settings.step_equilibration_sigma) * delta_h_ki**2
-                    elif settings.step_equilibration == StepEquilibrationMode.L2_RELAXED:
-                        eta_k = fe.prev_fe.sum_Lambda() * sum_Lambda_ki + fe.prev_fe.sum_Theta(
-                        ) * fe.sum_Theta()
-                        nu_k = 1
-                        for jjj in range(dims.n_theta):
-                            nu_k = nu_k * eta_k[jjj]
-                        nu_vector = vertcat(nu_vector, nu_k)
-                        self.objective = settings.rho_h * (nu_k) * delta_h_ki**2
+            for _, fe in enumerate(stage):
```
[the remark this answers is not in the document]
Comment (FreyJo) [reply nosnoc/nosnoc.py:728]: yeah, I was also happy to see it so compact :)

File: nosnoc/utils.py [hunk 3]
```diff
@@ -13,8 +12,8 @@ def validate(obj: object) -> bool:
     return
 
 
-def generate_butcher_tableu(n_s, irk_scheme):
-    if irk_scheme == IRKSchemes.RADAU_IIA:
+def generate_butcher_tableu(n_s, irk_string):
```
[the remark this answers is not in the document]
Comment (FreyJo) [reply nosnoc/utils.py:15]: Yes, I wanted to decouple.
We need to use strings in for getting tableaus from CasADi anyway.

[the remark this answers is not in the document]
Comment (FreyJo) [reply nosnoc/utils.py]: But, yes it is cleaner as you described, so I did it :+1:

**Comment by FreyJo** [comment]
Glad you like it! Should be all good now :)

**Review by apozharski — APPROVED** [review]
Looks good to me!

**Pull request merged**